src/iterators.jl

```diff
@@ -261,9 +310,71 @@ end
 @inline _getcache(fi::FaceIterator) = fi.fc
 @inline _getset(fi::FaceIterator) = fi.set
 
+"""
+    InterfaceIterator(grid::Grid, [topology::ExclusiveTopology])
+    InterfaceIterator(dh::AbstractDofHandler, [topology::ExclusiveTopology])
+    
+Create an `InterfaceIterator` to conveniently iterate over all the interfaces in a
+grid. The elements of the iterator are [`InterfaceCache`](@ref)s which are properly
+`reinit!`ialized. See [`InterfaceCache`](@ref) for more details.
+Looping over an `InterfaceIterator`, i.e.:
+```julia
+for ic in InterfaceIterator(grid, topology)
+    # ...
+end
+```
+is thus simply convenience for the following equivalent snippet for grids of dimensions > 1:
+```julia
+ic = InterfaceCache(grid, topology)
+for face in topology.face_skeleton
+    neighborhood = topology.face_face_neighbor[face[1], face[2]]
+    isempty(neighborhood) && continue
+    neighbor_face = neighborhood[1]
+    reinit!(ic, face, neighbor_face)
+    # ...
+end
+```
+!!! warning
+    `InterfaceIterator` is stateful and should not be used for things other than `for`-looping
+    (e.g. broadcasting over, or collecting the iterator may yield unexpected results).
+"""
+struct InterfaceIterator{Cache<:InterfaceCache}
+    cache::Cache
+    set::Set{NTuple{2, FaceIndex}}
+end
+
+function InterfaceIterator(gridordh::Union{Grid,AbstractDofHandler},
+                      topology::ExclusiveTopology = ExclusiveTopology(gridordh))
+    grid = gridordh isa Grid ? gridordh : get_grid(gridordh)
+    i = 1;
+    grid_dim = getdim(grid)
+    neighborhood = grid_dim == 1 ? topology.vertex_vertex_neighbor : topology.face_face_neighbor
+    interface_skeleton = Array{NTuple{2, FaceIndex}}(undef, count(face -> !isempty(neighborhood[face[1], face[2]]), faceskeleton(topology, grid)))
+    for face_a in faceskeleton(topology, grid)
+        !isempty(neighborhood[face_a[1], face_a[2]]) || continue
+        neighbor = neighborhood[face_a[1], face_a[2]].neighbor_info[]
+        face_b = neighbor isa VertexIndex ? FaceIndex(neighbor[1], neighbor[2]) : neighbor
+        interface_skeleton[i] = (face_a, face_b) #Assumes one neighbor only
+        i+=1
+    end
+    set = Set(interface_skeleton)
```

Comment:
I wonder if this can be done lazily. There doesn't seem to be a need to materialize the set here, right? Just loop over all of them and skip the ones which doesn't have a neighbor

Reply:
Like this?
https://github.com/Ferrite-FEM/Ferrite.jl/blob/7fe9140555b73137afbdd7188b808d1c468d98df/src/iterators.jl#L341-L371